File: 367.valid-perfect-square.js
/*
 * @lc app=leetcode id=367 lang=javascript
 *
 * [367] Valid Perfect Square
 */

// @lc code=start
/**
 * @param {number} num
 * @return {boolean}
 */
var isPerfectSquare = function (num) {
  if (num === 1) {
    return true;
  }

  let [left, right] = [0, num];
  while (left < right) {
    const mid = left + Math.trunc((right - left) / 2);
    if (mid < num / mid) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }
  return right * right === num;
};
// @lc code=end

File: 367.valid-perfect-square.go
/*
 * @lc app=leetcode id=367 lang=golang
 *
 * [367] Valid Perfect Square
 */

package leetcode

// @lc code=start
func isPerfectSquare(num int) bool {
	if num == 1 {
		return true
	}
	left, right := 0, num
	for left < right {
		mid := left + (right-left)/2
		if mid < num/mid {
			left = mid + 1
		} else {
			right = mid
		}
	}
	return right*right == num
}

// @lc code=end
